Fix incorrect local IP address sometimes put in QR code

wmergenthal opened the pull request:
When advertising its local IP address in a QR code, HearThis (HT) can sometimes show the wrong address. This is because it decides to use the wrong network interface. I observed this in my home setup three months ago. Windows 10 chose to use a virtual interface associated with my machine's VPN tunnel, and the tunnel's IP address got baked into the QR code for HearThisAndroid (HTA) to scan. This local IP address lived on a different subnet, which HTA's UDP response could not reach. Thus the desired HT/HTA connection and sync never occurred.

This changeset addresses the issue by determining the local IP in very similar fashion to a PR submitted yesterday for BloomDesktop. The sequence goes like this:
A) Examine the machine's network interfaces. Note that there could be several of these - my Win10 laptop has 11. Focus on the *active* Wi-Fi and Ethernet network interfaces and record their "interface metric" values.
B) Determine which interface the network stack will use for network traffic. Literature says the stack wants to use the interface with the lowest metric value. We want to prefer Wi-Fi over Ethernet. I have not seen a case where these criteria conflict, so the "winning" interface is selected thus:
    - IF a Wi-Fi interface is active, choose that
    - ELSE IF an Ethernet interface is active, choose that
    - ELSE declare 'none' and no local IP address will be declared
An interface is considered active if its interface metric is less than the initialized value of (2^31 - 1). On my machine at least, inactive interfaces all do maintain this initial value.
C) Note the winning interface's IP address. This is declared the Local IP address, to be put into the QR code.

Possible future improvement (unrelated to this PR's issue): modify the response from HTA to use TCP instead of UDP. If this were to be done, HT/HTA devices in separate subnets would at least have a fighting chance to connect. Routers are often configured to pass TCP traffic between subnets, but they pass UDP traffic much less often.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/hearthis/329)
<!-- Reviewable:end -->

tombogle commented:
I think it's probably worth adding a bullet point to the [release notes](https://github.com/sillsdev/hearthis/blob/master/DistFiles/ReleaseNotes.md) for this. I wouldn't make it long or overly technical. Maybe something like: "Improved connectivity with HearThis Android" or something along those lines.

wmergenthal commented:
These changes have been made:
1. Removed unused import 'NDesk.DBus'.
2. Reduced scope of the two InterfaceInfo structs so they can be local to GetInterfaceStackWillUse(). Their names are also modified to accommodate typical practice. GetInterfaceStackWillUse() now returns address strings directly so that enum CommTypeToExpect can be removed.
3. For string output to console, stop mixing string interpolation and concatenation.
4. GetIpForwardTable() does not need to be in try-catch block because it does not throw exceptions. Checking of its return value is improved a bit, however.
5. Part of GetMetricForInterface() is broken out into a new function GetRoutingTable(). This allows getting a routing table copy to be done once up front instead for each active network interface.
6. Some code comments and variable names are also modified for clarity.

wmergenthal commented:
Rather extensive refactor to implement change request in current HearThis pull request #329.
1. Adds class RoutingTableProxy as sole handler of routing table copy buffer.
2. Removes GetRoutingTable() and GetMetricForInterface() from class AndroidSynchronization.
3. Reimplements GetMetricForInterface() in the new proxy class.
4. Several smaller optimizations/improvements such as changing public items to private, renaming items for clarity, improving comments.

tombogle reviewed (commented):
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/hearthis/329)* status: all files reviewed, 1 unresolved discussion (waiting on @wmergenthal)



<!-- Sent from Reviewable.io -->
